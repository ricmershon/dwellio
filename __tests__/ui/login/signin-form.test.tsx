import React from 'react';
import { render, screen, fireEvent, setupCommonMocks } from '@/__tests__/test-utils';
import SignInForm from '@/ui/login/signin-form';

// Mock the Input component
jest.mock('@/ui/shared/input', () => {
    const MockInput = ({ id, name, type, label, placeholder, disabled, onClick, required }: any) => (
        <div data-testid={`input-${id}`}>
            <label htmlFor={id} className="mb-1 block text-sm text-gray-700">
                {label}
            </label>
            <input
                id={id}
                name={name}
                type={type}
                placeholder={placeholder}
                disabled={disabled}
                onClick={onClick}
                required={required}
                data-testid={`input-field-${id}`}
            />
        </div>
    );
    MockInput.displayName = 'MockInput';
    return MockInput;
});

describe('SignInForm', () => {
    const mockHandleSignIn = jest.fn();
    const mockHandleClearInfo = jest.fn();
    
    const defaultProps = {
        handleSignIn: mockHandleSignIn,
        isLoading: false,
        handleClearInfo: mockHandleClearInfo,
    };

    setupCommonMocks();

    describe('Component Structure', () => {
        it('should render all form fields correctly', () => {
            render(<SignInForm {...defaultProps} />);

            expect(screen.getByTestId('input-signin-email')).toBeInTheDocument();
            expect(screen.getByTestId('input-signin-password')).toBeInTheDocument();

            // Check labels
            expect(screen.getByText('Email Address *')).toBeInTheDocument();
            expect(screen.getByText('Password *')).toBeInTheDocument();
        });

        it('should render submit button with correct text', () => {
            render(<SignInForm {...defaultProps} />);

            const submitButton = screen.getByRole('button', { name: /sign in/i });
            expect(submitButton).toBeInTheDocument();
            expect(submitButton).toHaveClass('w-full', 'btn', 'btn-login-logout', 'mt-2');
        });
    });

    describe('Form Behavior', () => {
        it('should call handleSignIn on form submission', async () => {
            render(<SignInForm {...defaultProps} />);

            const form = document.querySelector('form');
            expect(form).toBeInTheDocument();
            fireEvent.submit(form!);

            expect(mockHandleSignIn).toHaveBeenCalledWith(expect.any(Object));
        });

        it('should call handleClearInfo when fields are clicked', () => {
            render(<SignInForm {...defaultProps} />);

            const emailField = screen.getByTestId('input-field-signin-email');
            const passwordField = screen.getByTestId('input-field-signin-password');

            fireEvent.click(emailField);
            fireEvent.click(passwordField);

            expect(mockHandleClearInfo).toHaveBeenCalledTimes(2);
        });
    });

    describe('Loading States', () => {
        it('should disable fields and button when isLoading is true', () => {
            render(<SignInForm {...defaultProps} isLoading={true} />);

            const emailField = screen.getByTestId('input-field-signin-email');
            const passwordField = screen.getByTestId('input-field-signin-password');
            const submitButton = screen.getByRole('button');

            expect(emailField).toBeDisabled();
            expect(passwordField).toBeDisabled();
            expect(submitButton).toBeDisabled();
        });

        it('should show loading text and spinner when loading', () => {
            render(<SignInForm {...defaultProps} isLoading={true} />);

            expect(screen.getByText('Please wait...')).toBeInTheDocument();
            expect(screen.getByRole('button')).toContainHTML('svg');
        });
    });

    describe('Field Configuration', () => {
        it('should configure email field correctly', () => {
            render(<SignInForm {...defaultProps} />);

            const emailField = screen.getByTestId('input-field-signin-email');
            
            expect(emailField).toHaveAttribute('type', 'email');
            expect(emailField).toHaveAttribute('name', 'email');
            expect(emailField).toHaveAttribute('placeholder', 'your@email.com');
            expect(emailField).toHaveAttribute('required');
        });

        it('should configure password field correctly', () => {
            render(<SignInForm {...defaultProps} />);

            const passwordField = screen.getByTestId('input-field-signin-password');
            
            expect(passwordField).toHaveAttribute('type', 'password');
            expect(passwordField).toHaveAttribute('name', 'password');
            expect(passwordField).toHaveAttribute('placeholder', 'Enter your password');
            expect(passwordField).toHaveAttribute('required');
        });
    });

    describe('Accessibility', () => {
        it('should have proper form structure', () => {
            render(<SignInForm {...defaultProps} />);

            const form = document.querySelector('form');
            expect(form).toBeInTheDocument();
            expect(form).toHaveClass('space-y-4');
        });

        it('should have proper button accessibility', () => {
            render(<SignInForm {...defaultProps} />);

            const button = screen.getByRole('button', { name: /sign in/i });
            expect(button).toHaveAttribute('type', 'submit');
        });

        it('should associate labels with form fields', () => {
            render(<SignInForm {...defaultProps} />);

            const emailLabel = screen.getByText('Email Address *');
            const passwordLabel = screen.getByText('Password *');

            expect(emailLabel).toHaveAttribute('for', 'signin-email');
            expect(passwordLabel).toHaveAttribute('for', 'signin-password');
        });
    });

    describe('Event Handling', () => {
        it('should prevent default form submission', async () => {
            render(<SignInForm {...defaultProps} />);

            const form = document.querySelector('form');
            expect(form).toBeInTheDocument();
            
            const submitEvent = new Event('submit', { bubbles: true, cancelable: true });
            const preventDefaultSpy = jest.spyOn(submitEvent, 'preventDefault');
            
            // This simulates the form submission without fireEvent
            form!.dispatchEvent(submitEvent);

            expect(mockHandleSignIn).toHaveBeenCalled();
        });
    });

    describe('Snapshots', () => {
        it('should match snapshot in default state', () => {
            const { container } = render(<SignInForm {...defaultProps} />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot in loading state', () => {
            const { container } = render(<SignInForm {...defaultProps} isLoading={true} />);
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});