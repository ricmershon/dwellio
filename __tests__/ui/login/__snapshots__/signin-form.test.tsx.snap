// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SignInForm Snapshots should match snapshot in default state 1`] = `
<form
  class="space-y-4"
>
  <div
    data-testid="input-signin-email"
  >
    <label
      class="mb-1 block text-sm text-gray-700"
      for="signin-email"
    >
      Email Address *
    </label>
    <input
      data-testid="input-field-signin-email"
      id="signin-email"
      name="email"
      placeholder="your@email.com"
      required=""
      type="email"
    />
  </div>
  <div
    data-testid="input-signin-password"
  >
    <label
      class="mb-1 block text-sm text-gray-700"
      for="signin-password"
    >
      Password *
    </label>
    <input
      data-testid="input-field-signin-password"
      id="signin-password"
      name="password"
      placeholder="Enter your password"
      required=""
      type="password"
    />
  </div>
  <button
    class="w-full btn btn-login-logout mt-2"
    type="submit"
  >
    Sign In
  </button>
</form>
`;

exports[`SignInForm Snapshots should match snapshot in loading state 1`] = `
<form
  class="space-y-4"
>
  <div
    data-testid="input-signin-email"
  >
    <label
      class="mb-1 block text-sm text-gray-700"
      for="signin-email"
    >
      Email Address *
    </label>
    <input
      data-testid="input-field-signin-email"
      disabled=""
      id="signin-email"
      name="email"
      placeholder="your@email.com"
      required=""
      type="email"
    />
  </div>
  <div
    data-testid="input-signin-password"
  >
    <label
      class="mb-1 block text-sm text-gray-700"
      for="signin-password"
    >
      Password *
    </label>
    <input
      data-testid="input-field-signin-password"
      disabled=""
      id="signin-password"
      name="password"
      placeholder="Enter your password"
      required=""
      type="password"
    />
  </div>
  <button
    class="w-full btn btn-login-logout mt-2"
    disabled=""
    type="submit"
  >
    <div
      class="flex items-center"
    >
      <svg
        class="animate-spin -ml-1 mr-2 h-4 w-4 text-white"
        fill="none"
        viewBox="0 0 24 24"
      >
        <circle
          class="opacity-25"
          cx="12"
          cy="12"
          r="10"
          stroke="currentColor"
          stroke-width="4"
        />
        <path
          class="opacity-75"
          d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
          fill="currentColor"
        />
      </svg>
      Please wait...
    </div>
  </button>
</form>
`;
